shell/platform/fuchsia/dart_runner/dart_component_controller_v2.cc

```diff
@@ -0,0 +1,489 @@
+  for (size_t i = url.length() - 1; i > 0; i--) {
+    if (url[i] == '/') {
+      return url.substr(i + 1, url.length() - 1);
+    }
+  }
+  return url;
+}
+
+// Find the name of the component.
+// fuchsia-pkg://fuchsia.com/hello_dart#meta/hello_dart.cm -> hello_dart
+std::string GetComponentNameFromUrl(const std::string& url) {
+  const std::string label = GetLabelFromUrl(url);
+  for (size_t i = 0; i < label.length(); ++i) {
+    if (label[i] == '.') {
+      return label.substr(0, i);
+    }
+  }
+  return label;
+}
+
+}  // namespace
+
+DartComponentControllerV2::DartComponentControllerV2(
+    fuchsia::component::runner::ComponentStartInfo start_info,
+    std::shared_ptr<sys::ServiceDirectory> runner_incoming_services,
+    fidl::InterfaceRequest<fuchsia::component::runner::ComponentController>
+        controller)
+    : loop_(new async::Loop(&kLoopConfig)),
+      label_(GetLabelFromUrl(start_info.resolved_url())),
+      url_(std::move(start_info.resolved_url())),
+      runner_incoming_services_(runner_incoming_services),
+      start_info_(std::move(start_info)),
+      binding_(this) {
+  // TODO(fxb/84537): This data path how we build Flutter applications in tree
+  // currently, but that may not always be the case. We should avoid making
+  // this assumption and let the CML file specify this data path instead.
+  const std::string component_name = GetComponentNameFromUrl(url_);
+  data_path_ = "pkg/data/" + component_name;
+
+  // TODO(akbiggs): In what situations is the controller not valid? Should
+  // we log an error?
+  if (controller.is_valid()) {
+    binding_.Bind(std::move(controller));
+    binding_.set_error_handler([this](zx_status_t status) { Kill(); });
+  }
+
+  zx_status_t idle_timer_status =
+      zx::timer::create(ZX_TIMER_SLACK_LATE, ZX_CLOCK_MONOTONIC, &idle_timer_);
+  if (idle_timer_status != ZX_OK) {
+    FX_LOGF(INFO, LOG_TAG, "Idle timer creation failed: %s",
+            zx_status_get_string(idle_timer_status));
+  } else {
+    idle_wait_.set_object(idle_timer_.get());
+    idle_wait_.set_trigger(ZX_TIMER_SIGNALED);
+    idle_wait_.Begin(async_get_default_dispatcher());
+  }
+}
+
+DartComponentControllerV2::~DartComponentControllerV2() {
+  if (namespace_) {
+    fdio_ns_destroy(namespace_);
+    namespace_ = nullptr;
+  }
+  close(stdout_fd_);
+  close(stderr_fd_);
+}
+
+bool DartComponentControllerV2::SetUp() {
+  // Name the thread after the url of the component being launched.
+  zx::thread::self()->set_property(ZX_PROP_NAME, label_.c_str(), label_.size());
+  Dart_SetThreadName(label_.c_str());
+
+  if (!CreateAndBindNamespace()) {
+    return false;
+  }
+
+  if (SetUpFromAppSnapshot()) {
+    FX_LOGF(INFO, LOG_TAG, "%s is running from an app snapshot", url_.c_str());
+  } else if (SetUpFromKernel()) {
+    FX_LOGF(INFO, LOG_TAG, "%s is running from kernel", url_.c_str());
+  } else {
+    FX_LOGF(ERROR, LOG_TAG,
+            "Could not find a program in %s. Was data specified"
```

Comment:
This error message technically isn't correct since we are not specifying data in the component manifest for v2 components.

Reply:
Switched to a more generic "setup failed" error message.